src/syntaxtree.c

```diff
@@ -31,50 +31,101 @@
 } ArgT;
 
 typedef struct {
     UInt1        tnum;
     CompileFuncT compile;
+    CodeFuncT    code;
     const Char * name;
-    UInt1        arity;
-    ArgT         args[4]; /* The maximum compiler arity is currently 4 */
+    UInt         arity;
+    ArgT args[MAX_ARITY]; /* The maximum compiler arity is currently 4 */
 } CompilerT;
 
 // We put compilers for statements and expressions into the same static array,
 // assuming that the set of their respective ids are disjoint.
 static const CompilerT Compilers[];
 #define COMPILER_ARITY(...) (sizeof((ArgT[]){ __VA_ARGS__ }) / sizeof(ArgT))
-#define COMPILER_GENERIC(tnum, compiler, ...)                                \
+#define COMPILER_GENERIC(tnum, compiler, coder, ...)                         \
     [tnum] = {                                                               \
-        tnum, compiler, #tnum, COMPILER_ARITY(__VA_ARGS__), { __VA_ARGS__ }  \
+        tnum,           compiler, coder, #tnum, COMPILER_ARITY(__VA_ARGS__), \
+        { __VA_ARGS__ }                                                      \
     }
 
-#define COMPILER(tnum, compiler) COMPILER_GENERIC(tnum, compiler)
+#define COMPILER(tnum, compiler, coder)                                      \
+    COMPILER_GENERIC(tnum, compiler, coder)
 
 #define COMPILER_(tnum, ...)                                                 \
-    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler, __VA_ARGS__)
+    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler,                        \
+                     SyntaxTreeDefaultCoder, __VA_ARGS__)
 
-#define ARG(name, func)                                                      \
+#define ARG(name, func, coder, isStat)                                       \
     {                                                                        \
-        name, func                                                           \
+        name, func, coder, isStat                                            \
     }
-#define ARG_(name) ARG(name, SyntaxTreeCompiler)
 
-#define ARGS(name) ARG(name, 0)
+#define ARG_EXPR(name, func, coder) ARG(name, func, coder, 0)
+// The corresponding ARG_STAT is not used
+
+#define ARG_STAT_(name)                                                      \
+    ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultStatCoder, 1)
+#define ARG_EXPR_(name)                                                      \
+    ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultExprCoder, 0)
 
+#define ARGS_STAT(name) ARG(name, 0, 0, 1)
+#define ARGS_EXPR(name) ARG(name, 0, 0, 0)
 
-static Obj typeStrings;
 
+static Obj typeStrings;    // plain list mapping statement tnums to strings
+                           // like "T_INTEXPR"
+static Obj typeRec;    // plain record mapping strings like "T_INTEXPR" to the
+                       // corresponding statement tnum
+
+static Obj ElmRecST(UInt1 tnum, Obj node, const char * name)
+{
+    UInt rnam = RNamName(name);
+    if (!IsbPRec(node, rnam)) {
```

Comment:
We never verify that the object we get is a precord, do we? So this would crash if I created malicious input (respectively: if I try to create a valid syntax tree object but screwed it up).

As with lists, we can either insert `RequirePRecord` tests everywhere (after creating such a function). Or we just use `ISB_REC`, `ELM_REC`, etc.

It might also be instructive to write malicious input just to see how it crashes and burns w/o these error checks. :-).